internal/states/encryption/encryptionconfig/config.go

```diff
@@ -0,0 +1,242 @@
+// this to specify a different encryption key or method when you want to read a state file from a different remote
+// storage than the backend for the current project.
+//
+// If the key is "terraform_remote_state.foo", its value sets/overrides the encryption configuration for
+//
+//	data "terraform_remote_state" "foo" {}
+//
+// For indexed resources, the format is "terraform_remote_state.foo[17]" or "terraform_remote_state.foo[key]"
+// (no quotes around the index).
+//
+// Use case:
+//
+// This is useful when you need to access information from another tofu environment. For example, you may want to
+// segment your tofu setup into multiple projects for both security and to optimize the run time. One such example
+// is Terragrunt. You may, for example, wish to obtain resources IDs for DNS entries from the project responsible for
+// DNS configuration for the purpose of setting up a web server.
+//
+// When using shared state and shared encryption in this fashion, the author of the other project must pay attention
+// to only expose information that is intended for the current environment in the state.
+type Key string
+
+const (
+	// KeyDefaultRemote contains the configuration key for default values applied only to remote state encryption,
+	// and only if the user provided no other configuration.
+	KeyDefaultRemote Key = "default_remote"
+	// KeyBackend contains the configuration key for overriding the default value when uploading the state to a
+	// remote backend.
+	KeyBackend Key = "backend"
+	// KeyStateFile contains the configuration key for encryption configuration for local state files. The default
+	// remote configuration is not applied to this encryption.
+	KeyStateFile Key = "statefile"
+	// KeyPlanFile contains the configuration key for encryption configuration for local plan files. The default
+	// remote configuration is not applied to this encryption.
+	KeyPlanFile Key = "planfile"
+)
+
+// UsesRemoteDefaults returns true if KeyDefaultRemote should be applied in addition to the current key.
+func (k Key) UsesRemoteDefaults() bool {
+	return k != KeyDefaultRemote && k != KeyStateFile && k != KeyPlanFile
+}
+
+// Validate checks if the key is a valid configuration key.
+func (k Key) Validate() error {
+	switch k {
+	case KeyDefaultRemote:
+		return nil
+	case KeyBackend:
+		return nil
+	case KeyPlanFile:
+		return nil
+	case KeyStateFile:
+		return nil
+	default:
+		if !k.IsRemoteDataSource() {
+			return fmt.Errorf(
+				"invalid encryption configuration key: %s (must be one of %s or contain a dot to specify a remote state data source)",
+				k,
+				strings.Join(
+					[]string{
+						string(KeyDefaultRemote), string(KeyBackend), string(KeyPlanFile), string(KeyStateFile),
+					}, ", ",
+				),
+			)
+		}
+		return nil
+	}
+}
+
+var remoteDatasourceRe = regexp.MustCompile(`^[a-z0-9_]+\.[a-z0-9_]+(|\[(\d+|[a-z0-9_]+)])$`)
```

Comment:
I'm not very familiar with the terraform_remote_state data source, but doesn't it always start with the string `terraform_remote_state` (I saw we don't support `tofu_remote_state`)? So why are we searching for a regex that contains a dot instead of a regex that fits `terraform_remote_state.`?

Reply:
I don't know if currently or in the future there may be another way to access the remote state. However, even if there is only this one way and we don't want to change it, the validation of that fact these start with `terraform_remote_state` shouldn't be done in this package, but rather in a central place in the codebase. This may be something for a future refactor PR.